moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,365 @@
+      return when {
+        methodSignatureIncludesJsonWriterAndJsonAdapter -> {
+          // void pointToJson(JsonWriter jsonWriter, Point point) {
+          // void pointToJson(JsonWriter jsonWriter, Point point, JsonAdapter<?> adapter, ...) {
+          val qualifierAnnotations = parameterAnnotations[1].jsonAnnotations
+          object : AdapterMethod(
+            adaptersOffset = 2,
+            type = parameterTypes[1],
+            parameterCount = parameterTypes.size,
+            annotations = qualifierAnnotations,
+            adapter = adapter,
+            method = method,
+            nullable = true
+          ) {
+            override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+              invoke(writer, value)
+            }
+          }
+        }
+        parameterTypes.size == 1 && returnType != Void.TYPE -> {
+          // List<Integer> pointToJson(Point point) {
+          val returnTypeAnnotations = method.jsonAnnotations
+          val qualifierAnnotations = parameterAnnotations[0].jsonAnnotations
+          val nullable = parameterAnnotations[0].hasNullable
+          object : AdapterMethod(
+            adaptersOffset = 1,
+            type = parameterTypes[0],
+            parameterCount = parameterTypes.size,
+            annotations = qualifierAnnotations,
+            adapter = adapter,
+            method = method,
+            nullable = nullable
+          ) {
+            private lateinit var delegate: JsonAdapter<Any>
+            override fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+              super.bind(moshi, factory)
+              delegate = if (Types.equals(parameterTypes[0], returnType) && qualifierAnnotations == returnTypeAnnotations) {
+                moshi.nextAdapter(factory, returnType, returnTypeAnnotations)
+              } else {
+                moshi.adapter(returnType, returnTypeAnnotations)
+              }
+            }
+
+            override fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+              val intermediate = invoke(value)
+              delegate.toJson(writer, intermediate)
+            }
+          }
+        }
+        else -> {
+          throw IllegalArgumentException(
+            "Unexpected signature for $method.\n@ToJson method signatures may have one of the following structures:\n    <any access modifier> void toJson(JsonWriter writer, T value) throws <any>;\n    <any access modifier> void toJson(JsonWriter writer, T value, JsonAdapter<any> delegate, <any more delegates>) throws <any>;\n    <any access modifier> R toJson(T value) throws <any>;\n"
+          )
+        }
+      }
+    }
+
+    /** Returns true if `parameterTypes[offset]` contains only JsonAdapters. */
+    private fun parametersAreJsonAdapters(offset: Int, parameterTypes: Array<Type>): Boolean {
+      for (i in offset until parameterTypes.size) {
+        if (parameterTypes[i] !is ParameterizedType) return false
```

Comment:
Save this into a local var so you can keep the smartcast

Reply:
https://github.com/square/moshi/pull/1467/commits/e23b50cb674287f73f484ab02f699ed456ae4ff3